File: NinePSharp.Messages.FSharp/messages/Rreadlink.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rreadlink =
    { Size: uint32
      Tag: uint16
      Target: string }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rreadlink
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rreadlink span &offset
            writeString this.Target span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let target = readString data &offset
        { Size = size; Tag = tag; Target = target }

File: NinePSharp/messages/Rreadlink.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Rreadlink : IMessage
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rreadlink;
    public ushort Tag { get; }
    public string Target { get; }

    public Rreadlink(uint size, ushort tag, string target)
    {
        Size = size;
        Tag = tag;
        Target = target;
    }

    public Rreadlink(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Target = data.ReadString(ref offset);
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Rreadlink);
        int offset = NinePConstants.HeaderSize;
        span.WriteString(Target, ref offset);
    }
}
